### ISSUE TITLE
3.2 release candidates discussion

### ISSUE THREAD

**Issue by nicm** [tmux/tmux#2280]
### Ask or discuss any quick questions here about 3.2 release candidates only please.

Bug reports please open a new issue with full information, see [CONTRIBUTING](https://github.com/tmux/tmux/blob/master/.github/CONTRIBUTING.md).

3.2-rc is a release candidate for 3.2.

**[The CHANGES file is here](https://raw.githubusercontent.com/tmux/tmux/3.2/CHANGES).**

---

There will be a **lot** of changes in this release, these are the most notable, for more see the CHANGES file:

- How data is sent to control mode clients has been completely revamped to both be more fair with multiple panes and to prevent huge amounts of data being backed up; this feature is used by default when tmux is attached with the latest iTerm2 nightly builds.

- Configuration file parsing has changed slightly: the contents of the new {} syntax introduced in 3.1 must now be valid tmux command syntax; and to allow formats to be annotated, strings given with quotes may now contain newlines (leading spaces and comments are stripped).

- A new customize mode available with `C-b C` (`C-b S-c`) which allows options and key bindings to be browsed and changed interactively.

- Support for extended keys offered by some terminals (xterm, mintty, iTerm2), see https://github.com/tmux/tmux/wiki/Modifier-Keys#extended-keys.

- A `pane-border-lines` option to change the characters used to draw the pane border separators.

- How UTF-8 data is stored has been rewritten to reduce memory use for characters in the BMP.

- The message log (`C-b ~`) has been changed to be per server instead of per client and to have some useful content.

- Client flags have been extended (`refresh-client -F` becomes `-f` and also added for `attach-session` and `new-session`), and a `read-only` flag added (same as `-r` to `attach-session`).

- A new `active-pane` client flag that if given allows a client to have its own active pane for each window rather than being tied to the server's active pane.

- Improved `command-prompt` completion including showing a menu of completions.

- All style options can now be formats, for example the default `pane-active-border-style` now changes colour depending on `pane_in_mode` and `synchronize-panes`.

- Search match count has come back in copy mode, as well as performance improvements and additional styles for marking of search terms.

- Try `$XDG_CONFIG_HOME/tmux/tmux.conf` as well as `~/.config/tmux/tmux.conf`.

- Window and pane hooks such as `window-layout-changed` and `pane-exited` are now window or pane options instead of session options.

- `copy-pipe` and friends can now be given with no arguments in which case they get the command from a new `copy-command` option. The default copy commands all now use `copy-pipe`. This makes it much easier to configure `copy-pipe` for all copy key bindings.

- Terminal detection and feature code has been rewritten. tmux now collects sets of related terminfo(5) capabilties together with sensible default values into named "terminal features" which makes them easier to configure via a new `terminal-features` option or `-T` flag to tmux itself.  tmux can detect and will automatically configure a number of common, modern terminal emulators.

- Support for per-client overlay popups with a new `display-popup` command. A popup is also used for an `e` key binding in buffer mode to edit a buffer using the editor given in the new `editor` option (filled in from the `EDITOR` or `VISUAL` environment variables by default).

- `DoubleClick` and `TripleClick` mouse key bindings have changed so that only one is fired. Behaviour similar to the old double-click key binding is now available as `SecondClick`. There are now default double- and triple-click bindings to copy a word or line, and a default middle-click binding to paste the top buffer.

- When a server socket is given with `-S`, it is now created with umask 177 rather than 117.

---

**Note the following known potentially-incompatible changes:**

- The {} syntax added in tmux 3.1 has changed so that the contents must follow the syntax of tmux commands.

- The `DoubleClick` key binding is not fired on the way to `TripleClick`; only one is fired. The new `SecondClick` key binding behaves similarly to the old `DoubleClick`.

**Comment by gnachman** [comment]
> The {} syntax added in tmux 3.1 has changed so that the parser that the contents be valid tmux commands.

Having trouble parsing this sentence.

**Comment by nicm** [comment]
> Having trouble parsing this sentence.

Yes I have changed that text now although I am not sure it is better.

The change is that previously you could put anything you liked in {} and it would be taken as a literal string. But that meant a separate parser for {} which was a mess - now it uses the same parser as the rest of the configuration file. So what appears in {} must be valid configuration file syntax. It doesn't have to actually be tmux commands, it just has to *look* like them.

The main effect of this are that you can no longer use {} to contain sections of shell script; I have code to allow this by marking {} with a modifier saying "this is shell script" but I am not going to to apply it or decide a way forward until I hear how much effect the change to {} has on real configuration files. {} has only been present at all for one release after all.

**Comment by lhoenig** [comment]
Why is the tag for 3.2-rc called "3.2"? Is this intended?

**Comment by mmjo** [comment]
> Why is the tag for 3.2-rc called "3.2"? Is this intended?

That confused me

**Comment by kevinhwang91** [comment]
Could set-mark be persistent? I mean if I have set-mark before, and then enter copy-mode again, press M-X will go back the mark I set before.

**Comment by nicm** [comment]
It could be done but it would be fiddly to keep it right if the pane scrolled.

**Comment by kevinhwang91** [comment]
> * Add a copy-mode -H flag to hide the position marker in the top right.

How can I toggle this flag on the fly? Sometime the marker will override the top right content, I must scroll up and down to solve this issue. Or add a pane-style?

**Comment by Tracerneo** [comment]
> A new `active-pane` client flag that if given allows a client to have its own active pane for each window rather than being tied to the server's active pane.

Wouldn't `multiseat` or alternatively `client-pane` be a better name for `active-pane` flag?

**Comment by fullstopslash** [comment]
Glad to see XDG_CONFIG_HOME finally being adopted.

**Comment by kevinhwang91** [comment]
> * Add a mark to copy mode, this is set with the set-mark command (bound to X)
  and appears with the entire line shown using copy-mode-mark-style and the
  marked character in reverse. The jump-to-mark command (bound to M-x) swaps
  the mark and the cursor positions.

How can I prevent `jump-to-mark` from swapping?

**Comment by mhanberg** [comment]
Thanks for your work on tmux, i've used it everyday for several years and I just starting using the 3.2-rc release.

I'm having some trouble getting my terminal colors to work in 3.2 the same way that I had them working in the last release.

In order to get italics working previously, I had the following configuration and terminfo (I applied the terminfo using `tic -x tmux-terminfo.txt`). https://gist.github.com/mhanberg/cb7db5c40111dd6b2776e7512a30040b.

With 3.2-rc, the italics don't seem to work any longer, as well as some other color changes. I attempted to fix this by using the `terminal-features` configuration option, but it doesn't seem to have an option for italics.

One color oddity is that the borders in my `fzf.vim`, so I apologize if the problem ends up being with it. The gray border here is the oddity.
![image](https://user-images.githubusercontent.com/5523984/92623373-22bb4000-f294-11ea-9bb2-51c228ca7a04.png)

It should look like this
![image](https://user-images.githubusercontent.com/5523984/92622179-c277ce80-f292-11ea-87d2-eb85aab18129.png)

So I guess my question is, what is the recommended way to get true colors and italics working?

---

I have also noticed that copying (⌘+C in the terminal Alacritty) no longer works for me.

Thoughts? Thanks

**Comment by nicm** [comment]
Please open an issue with full config and logs and I will look at it next week.

**Comment by dkarter** [comment]
@mhanberg @nicm I'm having a similar issue with colors (in NeoVim), also with borders becoming dashed when they were not in 3.1 (see top border in After picture).

| Before  | After |
|---------|-------|
|  <img width="2032" alt="before" src="https://user-images.githubusercontent.com/551858/93040817-846f0780-f610-11ea-8fb4-c390b36f4639.png">  |  <img width="2672" alt="image" src="https://user-images.githubusercontent.com/551858/93040775-64d7df00-f610-11ea-995e-e87f8440f08d.png"> |

It looks like true color is turned off for panes with neovim? I don't know how else to explain this it, because colors look fine in regular panes.

---

@mhanberg For the cursor shapes in neovim I have been using this trick for some time, it is still working for me in 3.2, maybe give it a shot and see if it helps:

```tmux
# =================================================
# For NeoVim
# =================================================

  # https://github.com/neovim/neovim/wiki/FAQ#cursor-shape-doesnt-change-in-tmux
  set -g -a terminal-overrides ',*:Ss=\E[%p1%d q:Se=\E[2 q'
```

**Comment by dkarter** [comment]
I'm seeing some weird and unworkable behavior using [powerlevel10k](https://github.com/romkatv/powerlevel10k)
![2020-09-14 10 37 58](https://user-images.githubusercontent.com/551858/93108298-40175200-f678-11ea-8a26-9a2389afae2a.gif)

Is anyone else experiencing this?

**Comment by timtyrrell** [comment]
Per all the tmux/kitty truecolor discussion, this recent PR fixed it for me 🥇 https://github.com/tmux/tmux/issues/2418

**Comment by cevhyruz** [comment]
> * The {} syntax added in tmux 3.1 has changed so that the contents must follow the syntax of tmux commands.

I liked the way it was before in 3.1 as I can just  set an alias this way which is
I think more readable than throwing a newline ('\') for every line
while escaping braces.

PS: This also retains syntax highlighting which makes it more easier to read.

```
set-option -s command-alias[12] left={
  if-shell "$IS_VIM" {
    %if "#{m/r:(copy|view)-mode,#{pane_mode}}"
      select-pane -L
    %else
      send-keys C-h
    %endif
  } {
    select-pane -L
  }
} 
```
Is this possible to not be deprecated or at least only allow this for command-alias?







Just a now

**Comment by cevhyruz** [comment]
> * The {} syntax added in tmux 3.1 has changed so that the contents must follow the syntax of tmux commands.

Since the contents of command-alias is a valid configuration syntax isn't it better to just let it use {} for a more readable script?

```
set-option -s command-alias[10] up={
  if-shell "$IS_VIM || $IS_FZF" {
    %if "#{m/r:(copy|view)-mode,#{pane_mode}}"
      select-pane -U
    %else
      send-keys C-k
    %endif
  } { select-pane -U }
}
```

**Comment by nicm** [comment]
I wrote a change to allow {} to be marked as shell and used for shell
sections as before but nobody seemed much interested, I'll dig it out next
week and you can see if it works for this.

Alternatively it may be worth a special case for command-alias because it
is a command.

**Comment by cevhyruz** [comment]
> I wrote a change to allow {} to be marked as shell and used for shell sections as before but nobody seemed much interested

Haven't really tried scripting with this new change in {} syntax but I assume It's alright as it doesn't break too much of my config file (only the command aliases part).

I really think the command-alias should really have a special case to allow usage of {} (like in the snippet above)
as it is the only one (I think) that has potential risk of being convoluted with nested syntax.

Anyway I'll wait for the the patch  and give it a try.
Thanks for the the hardwork in tmux :+1:

**Comment by nicm** [comment]
I don't think we can have both the new way and the old way, because for example you are using `%if` and expecting it to be stored in the option value string rather than expanded at parse time. But I'm not sure why you are mixing `if-shell` and `%if` anyway, isn't it just the same doing this:
~~~~
set-option -g command-alias[10] up='
  if-shell "$IS_VIM || $IS_FZF" {
     if-shell -F "#{m/r:(copy|view)-mode,#{pane_mode}}" {
         select-pane -U
   } {
      send-keys C-k
   } } { select-pane -U }'
~~~~
One of the reasons it was changed was so that `%if` would work the same inside and outside `{}`. Not that I'm not particularly happy about how it works now, but I wasn't particularly happy about it before either and it is tidier and more consistent now...

**Comment by cevhyruz** [comment]
> But I'm not sure why you are mixing if-shell and %if anyway, isn't it just the same doing this:

You're right. I completely forgot about the format expansion.
This should be enough for  me now. I still want to try the patch though :)

**Comment by nicm** [comment]
The patch won't work for you, and I don't think it is that useful all round. So either we keep it as it is or revert to the old way, but since your example is pretty easy to change I don't think it is an argument either way :-).

**Comment by Jecvay** [comment]
after use tmux-popup in with fzf.vim

vim in tmux cannot use mouse.

**Comment by carlocab** [comment]
@Jecvay I think this is related to https://github.com/tmux/tmux/issues/2458. It should be fixed in master.

**Comment by dsedivec** [comment]
> * Configuration file parsing has changed slightly: the contents of the new {} syntax introduced in 3.1 must now be valid tmux command syntax; and to allow formats to be annotated, strings given with quotes may now contain newlines (leading spaces and comments are stripped).

I'm having some trouble understanding how parsing works now.  I've had this in my `tmux.conf`:

```
bind o run-shell {
	num_panes $(tmux display -p '##{window_panes}')
	if [ "$num_panes" -le 1 ]; then
		tmux display "No other pane to switch to"
	elif [ "$num_panes" -le 2 ]; then
		tmux last-pane
	else
		tmux display-panes -d 0
	fi
}
```

Now when I do prefix+o I get an error (the error is all on one line but I've wrapped it here for ease of reading):

```
'display -p "#{window_panes})" ; if [ "$(tmux" -le 1 ] ;
then ; tmux display "No other pane to switch to" ;
elif [ "$(tmux" -le 2 ] ; then ; tmux last-pane ; 
else ; tmux display-panes -d 0 ; fi' returned 2 
```

Could anyone help me understand how this needs to be changed for 3.2?  I'm running bfdc4373d71895a5f454834da4b0bd92ab964dd4 on macOS 10.15.7, my shell is Bash 5.0.17 (from MacPorts).  Thanks!

**Comment by nicm** [comment]
You can do it like this:
~~~~
bind o run-shell "
         num_panes $(tmux display -p '##{window_panes}'); \
         if [ \"$num_panes\" -le 1 ]; then \
                 tmux display \"No other pane to switch to\"; \
         elif [ \"$num_panes\" -le 2 ]; then \
                 tmux last-pane; \
         else \
                 tmux display-panes -d 0; \
         fi
"
~~~~
But I am not too happy about this change because it does make it hard to include shell scripts so I might change it back at some point or change it in some way.

**Comment by nicm** [comment]
I intended it to leave the newlines, I've just fixed that so if you wait a few hours for that change to sync you can do:
~~~~
bind o run-shell "
         num_panes $(tmux display -p '##{window_panes}')
         if [ \"$num_panes\" -le 1 ]; then
                 tmux display \"No other pane to switch to\"
         elif [ \"$num_panes\" -le 2 ]; then
                 tmux last-pane
         else
                 tmux display-panes -d 0
         fi
"

**Comment by laktak** [comment]
@nicm I think [this](https://github.com/tmux/tmux/wiki/Contributing) is no longer true (at least for the last few releases):

> tmux currently sees a new release approximately every six months - the same schedule as OpenBSD, around May and October. This means that the GitHub releases contain roughly the same code as OpenBSD releases (but not necessarily exactly the same).

Do you have a goal for the 3.2 release date? Thanks!

**Comment by nicm** [comment]
It is true for the last few, but there were a lot of changes in the last release and I have been busy so I am skipping this one and will release 3.2 when the next OpenBSD release comes probably around May.

**Comment by kevinhwang91** [comment]
It's so pretty to hear it, so many plugins are waiting for 3.2, added a lot of features. Nevermind, go busy with what you think is important, too many unexpected things this year.

**Comment by nicm** [comment]
Yes but a lot of changes means a lot of potential problems and I don't have time for them at the moment and since I don't have time to add lots more new features now, another few months to let people test is helpful.

**Comment by carlocab** [comment]
Please take your time, @nicm. I'm sure a lot of people greatly appreciate the work you do. (At the very least, I know I do.)

**Comment by kevinhwang91** [comment]
> Yes but a lot of changes means a lot of potential problems and I don't have time for them at the moment and since I don't have time to add lots more new features now, another few months to let people test is helpful.

I can't live without 3.2 version. I would test the master branch and post the issue if I found it.

**Comment by laktak** [comment]
@nicm Thanks for letting us know!

Now that the release is still a few months off I guess more people might be willing to install the git version. For those on arch, it's in the AUR: https://aur.archlinux.org/packages/tmux-git

**Comment by dsedivec** [comment]
> I've just fixed that so if you wait a few hours for that change to sync you can do:

Thanks so much!  I built latest master from just now and I tried [what you posted there](https://github.com/tmux/tmux/issues/2280#issuecomment-736399579), but it didn't work: I no longer got an error, but it seems like the `$`s in `$num_panes` were being expanded too early.  Does tmux itself expand "$foo"?

In any case, changing from double quotes to single quotes seems to have worked:

```
bind o run-shell '
	num_panes=$(tmux display -p "##{window_panes}")
	if [ "$num_panes" -le 1 ]; then
		tmux display "No other pane to switch to"
	elif [ "$num_panes" -le 2 ]; then
		tmux last-pane
	else
		tmux display-panes -d 0
	fi
'
```

(Reading this again, I'm a bit surprised it worked, I would have guessed that `##{window_panes}` would have been left as-is since I used single quotes.  ¯\\_(ツ)_/¯)

**Comment by nicm** [comment]
Oh yeah the config parser will expand $x inside double quotes.

**Comment by damnskippy** [comment]
I was trying to convert your [sweet recipe ](https://github.com/tmux/tmux/wiki/Recipes#c-doubleclick-to-open-emacs1
) to split-window instead of a popup but couldn't get it to do what I wanted - I guess the issue I'm facing comes down to passing tmux variables as arguments in the shell command, I'm probably not doing something correctly, and could use some help.  Thanks.

**Comment by nicm** [comment]
It is a pain with escaping and quoting, this works for me but TBH by this
point I would probably have just written a shell script rather than doing
it entirely in .tmux.conf:

bind -n C-DoubleClick1Pane if -F '#{m/r:^[^:]*:[0-9]+:,#{mouse_word}}' {
        run '
            tmux splitw -c "#{pane_current_path}" \
                "emacs `echo #{mouse_word}|awk -F: \"{print
\\"+\\"\\$2,\\$1}\"`"
        '
} {
        run '
            tmux splitw -c "#{pane_current_path}" \
                "emacs \"#{mouse_word}\""
        '
}

**Comment by damnskippy** [comment]
Thank you.

**Comment by systemmonkey42** [comment]
I would like to propose the following;

- Support `#{...}` expansion **everywhere**.   Avoids the need to use `run-shell tmux command #{..}` just to trigger expansion
 
- Support capturing output of command like `if-shell` and `display-popup` as `#{....}` style fields (eg, `#{arg0}` for the entire line, and `#{argN}` for the Nth word.
Eg
`if-shell "hostname" "display-message 'Hostname is #{arg1}'"` 

I currently use something like;

`tmux display-popup -EKR 'ls|tmux set-option @cc "$(fzf)"' \; display-message -p "#{@cc}"`

which opens the popup with a file list, allows one to be selected, with the result being placed into `@cc` and
output to the console when done. (This could also be done with temp-files)

Some way to more easily capture output from a shell to `#{...}` variable would be very useful.

Thanks

**Comment by nicm** [comment]
Possibly but:

1) Changing everything to support #{} is going to break existing scripts. Both some that use # and #{} for another purpose and some that expect them to be expanded at a later point.

2) It is probably some work to add format expansion everywhere, not least because it is not clear where it is missing.

I think a starting point would be to make a list of everything that currently does not expand. Then we can decide which can be changed immediately without any impact and which need a backwards compatible way to do it (such as a -F flag). Do you want to do that?

Putting command arguments in format variables is already done for hooks (look at cmdq_insert_hook) and should be easy enough to do for other commands - when there is a command it is given to format_create in the cmdq_item.

**Comment by systemmonkey42** [comment]
> Changing everything to support #{} is going to break existing scripts. Both some that use # and #{} for another purpose and some that expect them to be expanded at a later point.

Probably more a 4.0 feature then..

I mentioned in another thread that if `-s` and `-t` *always* supported format expansion implicitly, that would solve most use cases.

Can than be done safely without breaking existing code?

**Comment by nicm** [comment]
What use cases need targets to be expanded?

**Comment by systemmonkey42** [comment]
> What use cases need targets to be expanded?

For me its usually managing panes and hooks.
An obvious case being the example in #2513 where in a `pane-died` hook, we need to use 
```
run-shell "tmux kill-pane -t #{pane_id}"
```
when a simple `kill-pane -t #{pane_id}` should work.

**Comment by systemmonkey42** [comment]
> It's so pretty to hear it, so many plugins are waiting for 3.2, added a lot of features. Nevermind, go busy with what you think is important, too many unexpected things this year.

Where can I find these plugins you refer to?

Most plugins I've tried are so old they need rewriting to work with TMUX > 1.7, so I've ended up ignore plugins entirely.

**Comment by nicm** [comment]
You shouldn't need to do `-t #{pane_id}` without `run-shell`, it will be the default target. With `run-shell` it may be wise but you can already have formats replaced.

**Comment by laktak** [comment]
@systemmonkey42 take a look at https://github.com/rothgar/awesome-tmux for a list. For example [extrakto](https://github.com/laktak/extrakto) will work with popups if you have tmux >= 3.2

**Comment by ScrumpyJack** [comment]
Excited about the bracketed paste by default, which will make pasting into a REPL so much better

**Comment by laktak** [comment]
If I have `set-option -g set-clipboard on` (to copy from inside a ssh session with OSC52) and use `copy-pipe`, the text will be put on the system clipboard.

I was expecting to `copy-pipe` to only pass the text to my command.

Not sure if this is a bug because the manpage is not 100% clear ("the copied text is piped"):

> Copy commands may take an optional buffer prefix argument which is used to generate the buffer name (the default is ‘buffer’ so buffers are named ‘buffer0’, ‘buffer1’ and so on).  Pipe commands take a command argument which is the command to which the copied text is piped.  The ‘-and-cancel’ variants of some commands exit copy mode after they have completed (for copy commands) or when the cursor reaches the bottom (for scrolling commands).  ‘-no-clear’ variants do not clear the selection.

If it is not a bug - am I doing something wrong? Using the latest version on git.

**Comment by nicm** [comment]
`set-clipboard` applies to all the copy commands.

**Comment by laktak** [comment]
OK, thanks. In that case, would it be possible to add a `pipe` command or is there another way to solve this?

**Comment by nicm** [comment]
Turn off set-clipboard before you run your command and turn it on afterwards? You can do that in the key binding.

**Comment by nicm** [comment]
I would be OK with a `pipe` or `pipe-selection` command also if you want to add it :-).

**Comment by taybart** [comment]
@nicm Did the status formatter change? I am not getting the correct time when I add something like `[#(TZ="Etc/UTC" date +%%H)]` to my status. It is giving me back the system time instead of UTC. This seems to work on 3.1c though.

**Comment by nicm** [comment]
Yes, when #() is expanded again it should not expand time formats again (or I think further #() for that matter). Please try this: [x.diff.txt](https://github.com/tmux/tmux/files/5945875/x.diff.txt)

**Comment by taybart** [comment]
After poking around a bit (I had not used flags before), I am not sure where it would go. I accidentally copied the brackets `[]` Those are just literal brackets. The prompt `#(TZ="Etc/UTC" date +%%H)` should just be a shell command right? Does the hour format need the expand time flag?

**Comment by nicm** [comment]
I don't know what you mean? Try to apply the patch I gave you with `git apply x.diff.txt` then rebuild and install tmux and see if the problem is fixed.

**Comment by taybart** [comment]
Sorry about that! That patch did do the business. Is that an issue with 3.2 or my status format?

**Comment by taybart** [comment]
Thank you so much for all your work @nicm!

**Comment by joshmedeski** [comment]
I'm looking forward to running `apt install tmux` and get popup support on Ubuntu! Still learning how to build it from source. Let me know if you need help with testing.

**Comment by aaronjensen** [comment]
It's unclear if this is a 3.2 thing and/or if it's even meant to be fixed or not, but it seems like search wrapping (i.e. #1935) is partially working but doesn't work properly with wildcards.

For example:

```
echo aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab
```

Ensure it's wrapping by sizing your pane accordingly.

Then search backward for `a*b`.

The first match will be only the second half of the wrapped line. The second match will be only the first half of the wrapped line. Neither of these will do the right thing when `copy-pipe` is used.

I'd be happy to open another issue for this if this isn't the right place.

**Comment by nicm** [comment]
Is this master? Please open an issue.

**Comment by aaronjensen** [comment]
> Is this master? Please open an issue.

Yep, #2570

**Comment by damnskippy** [comment]
Just curious if this is user error on my part (likely) or if this is to be expected.  With regards to be able to change the terminal colors on the fly, the [base16 scripts](https://github.com/chriskempson/base16-shell/tree/master/scripts) are super useful, at least I find.  However sourcing the script while inside tmux doesn't quite affect the tmux color palette it seems, or you can see a hint of the new terminal colors around the window borders and not until you exit out of the tmux session.  Sourcing the same script first and then starting the tmux session works fine - tmux honors the sourced color palette.  Is this to be expected, and is there a way to work around this so as to be able to source the color script and effect the tmux palette on the fly?  Thanks.

**Comment by systemmonkey42** [comment]
@damnskippy Have you tried refreshing the client?   I suspect the escape codes change the colours directly in the terminal, but tmux will not redraw using the new colours because it doesn't know anthing has changed.

`tmux refresh-client`  (not sure of the default binding, possibly <kbd>Ctrl-B</kbd> followed by <kbd>l</kbd> or <kbd>Ctrl-L</kbd>)

Another option is to detach from the session, and re-attach, or even resize the window.

**Comment by damnskippy** [comment]
Indeed, I was wondering if there's a way to refresh the tmux's color palette after terminal colors have changed.

I tried both refresh-client and detach/re-attach and neither worked unfortunately.

**Comment by damnskippy** [comment]
Never mind, this was due to a bug in my own environment somewhere, I'll narrow this down.  Thanks.

**Comment by vimpostor** [comment]
In case someone else stumbles upon this: I had `set -g status-bg default` in my config to let tmux have the same background in the statusline. This doesn't seem to work anymore, I bisected the bad commit to be 1f8256fc508bbee24b53fcd588ebf74653d69dfa

However I don't know whether this qualifies as an issue or is intended, as I was able to fix it by making the following change to my config:

```diff
-set -g status-bg default
+set -g status-style bg=default
```

**Comment by laktak** [comment]
If I start tmux with
```
bind-key M-v run-shell "\1"
```
in the config it will report an error but if I reload the config while tmux is running the service will terminate. 

Is this a known bug or should I file an issue? (latest git)

**Comment by nicm** [comment]
Open an issue please if you can reproduce with master or 3.2-rc4.

**Comment by nicm** [comment]
Moving to https://github.com/tmux/tmux/issues/2592.
